library/ssl_tls13_client.c

```diff
@@ -2277,19 +2365,70 @@ static int ssl_certificate_request_coordinate( mbedtls_ssl_context* ssl )
         return( SSL_CERTIFICATE_REQUEST_SKIP );
     }
 
-#if !defined(MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED)
-    ( ( void )ret );
-    MBEDTLS_SSL_DEBUG_MSG( 1, ( "should never happen" ) );
-    return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
-#else
-    if( ssl->keep_current_message == 0 )
+    MBEDTLS_SSL_PROC_CHK( mbedtls_mps_read( &ssl->mps.l4 ) );
+    if( ret != MBEDTLS_MPS_MSG_HS )
     {
-        if( ( ret = mbedtls_ssl_read_record( ssl, 0 ) ) != 0 )
-        {
-            MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_ssl_read_record", ret );
-            return( ret );
-        }
+        MBEDTLS_SSL_DEBUG_MSG( 1, ( "bad certificate request message" ) );
+        SSL_PEND_FATAL_ALERT( MBEDTLS_SSL_ALERT_MSG_UNEXPECTED_MESSAGE );
+        return( MBEDTLS_ERR_SSL_UNEXPECTED_MESSAGE );
+    }
+
+    MBEDTLS_SSL_PROC_CHK( mbedtls_mps_read_handshake( &ssl->mps.l4, &msg ) );
+
+    if( msg.type == MBEDTLS_SSL_HS_CERTIFICATE_REQUEST )
+        return( SSL_CERTIFICATE_REQUEST_EXPECT_REQUEST );
+
+    return( SSL_CERTIFICATE_REQUEST_SKIP );
+
+cleanup:
+    return( ret);
+}
+
+static int ssl_certificate_request_fetch( mbedtls_ssl_context* ssl,
+                                          mbedtls_mps_handshake_in *msg )
+{
+    int ret;
+    MBEDTLS_SSL_PROC_CHK( mbedtls_mps_read( &ssl->mps.l4 ) );
+    if( ret != MBEDTLS_MPS_MSG_HS )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 1, ( "bad certificate request message" ) );
+        SSL_PEND_FATAL_ALERT( MBEDTLS_SSL_ALERT_MSG_UNEXPECTED_MESSAGE );
+        return( MBEDTLS_ERR_SSL_UNEXPECTED_MESSAGE );
+    }
+
+    MBEDTLS_SSL_PROC_CHK( mbedtls_mps_read_handshake( &ssl->mps.l4, msg ) );
+
+    if( msg->type != MBEDTLS_SSL_HS_CERTIFICATE_REQUEST )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 1, ( "bad certificate request message" ) );
+        SSL_PEND_FATAL_ALERT( MBEDTLS_SSL_ALERT_MSG_UNEXPECTED_MESSAGE );
+        return( MBEDTLS_ERR_SSL_UNEXPECTED_MESSAGE );
+    }
+
+cleanup:
+
+    return( ret );
+}
+
+#else /* MBEDTLS_SSL_USE_MPS */
+
+static int ssl_certificate_request_coordinate( mbedtls_ssl_context* ssl )
+{
+    int ret;
+
+    if( ssl->session_negotiate->key_exchange == MBEDTLS_KEY_EXCHANGE_PSK ||
+        ssl->session_negotiate->key_exchange == MBEDTLS_KEY_EXCHANGE_ECDHE_PSK )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 3, ( "<= skip parse certificate request" ) );
+        return( SSL_CERTIFICATE_REQUEST_SKIP );
+    }
+
+    if( ( ret = mbedtls_ssl_read_record( ssl, 0 ) ) != 0 )
+    {
+        MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_ssl_read_record", ret );
+        return( ret );
     }
+    ssl->keep_current_message = 1;
```

Comment:
Similar to comments in L2930, We probably should only set `keep_current_message` when the message (`ssl->in_msg[0]`) is not `MBEDTLS_SSL_HS_CERTIFICATE_REQUEST`

Reply:
This depends on whether the case `SSL_CERTIFICATE_EXPECT_REQUEST` is treated as a standalone handler which calls `mbedtls_ssl_read_record()` as the first step. If so, we need the `keep_current_message`. If not, we can leave it.

The current code comes at the cost of one call to `mbedtls_ssl_read_record()` which is essentially a no-op, but it keeps the structure of the handler independent of any prior coordination logic, which I think is beneficial for code clarity.